### PR TITLE
Implement release statuses

### PR DESCRIPTION
**Pull request by casperisfine**
This allows to publish commit statuses based on the deployed state:

Example from a `shipit.yml`:
```yaml
release:
  context: shipit/pre-production
  delay: 5m
```

capture d ecran 2018-09-11 a 12 45 29

This will:
  - publish a `pending` status on the deployed commit when a deploy is triggered.
  - If the deploy fails, an `error` status will be published
  - If `delay` is `0` a `success` status will be published as soon as the deploy succeed.
  - If `delay` is `null`, no status will be published on deploy success.
  - If `delay` is positive, a `success` status will be published after this amount of time, unless another one was published in the meantime.
  - If the deploy is rolled back, a `failure` status will be published.
  - Users can manually flip the status to either `failure` or `success`.

All this primitives allow to daisy chain stacks. For example you can have a `production` stack require a `shipit/staging` status. Users will have to deploy in staging first, and validate the release there first.

NB: This feature is still somewhat experimental, I'd rather not document it until we've ironed it out in production for a while as it's very much subject to changes.

### CODE REVIEW

File: test/models/release_statuses_test.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        description: 'Deploy started',
+      ).returns(resource(id: 42))
+
+      @status = shipit_release_statuses(:to_be_created)
+      assert_nil @status.github_id
+      @status.create_status_on_github!
+      assert_equal 42, @status.github_id
+    end
+
+    test "#create_status_on_github! does nothing if the github_id is alreayd recorded" do
```
Comment (DazWorrall) [inline test/models/release_statuses_test.rb]: nit: typo `alreayd`

**Comment by casperisfine** [comment]
> Is the UI the same as you demoed to us recently?

Yes. I did some very minor adjustments. I'm not super happy with it, but I hope our users will have suggestions for improvements.

> Do you expect an external analysis engine to hit a Shipit API if it determines a release is bad,

I expect it to trigger the rollback via the API, which will mark the release as unhealthy.

> During the recent test the release status icon was unexpectedly exposed to users, was that fixed?

Yes.

File: app/models/shipit/deploy.rb [hunk 2]
```diff
@@ -143,6 +145,21 @@ def confirmed?
       confirmations.abs >= CONFIRMATIONS_REQUIRED
     end
 
+    delegate :last_release_status, to: :until_commit
+    def append_release_status(state, description, user: self.user)
+      status = until_commit.create_release_status!(
+        state,
+        user: user.presence,
+        target_url: permalink,
```
[the remark this answers is not in the document]
Comment (casperisfine) [reply app/models/shipit/deploy.rb]: Yes good point. We'll have to add an API for validation from external tools.

**Comment by wvanbergen** [comment]
Maybe we can ask the dev ATC for feedback about the UI? Is there an experimental stack somewhere that we can point people to?

**Comment by casperisfine** [comment]
> Is there an experimental stack somewhere that we can point people to?

I rolled the preview back. I can bring it back if you wish. But yes, good idea.

**Comment by wvanbergen** [comment]
I wonder whether one CI status is going to cut it. It would be nice if we could have separate systems saying something about the quality of a release.
On the other hand, they could also publish their own CI status, but then we would have to have a way for shipit to pick up on these, and action them (e.g. a rollback)

**Comment by dwradcliffe** [comment]
I haven't reviewed the code but the feature description sounds good to me 👍 Looking forward to seeing how it works in real life.

**Pull request merged**